### ISSUE TITLE
Introduce materialization support for tables

### ISSUE THREAD

**Issue by derekjn** [confluentinc/ksql#5652]
Users often want to issue pull queries against regular tables. But this is not currently possible because only aggregate tables are materialized. The `latest_by_offset` aggregate can be used here as a workaround, but this option does not respect tombstones.

We should just give users the option to force the materialization of any table. As mentioned on #5640, it could work like so:

```sql
CREATE TABLE t0 (x INTEGER) WITH (materialize=true, ...);
CREATE TABLE t1 WITH (materialize=true, ...) AS SELECT ...
```
This table would be materialized into RocksDB, and thus able to serve pull queries while respecting tombstones.

**Comment by big-andy-coates** [comment]
And it's not just `CREATE TABLE` that needs this. `CREATE TABLE AS SELECT` also needs this in some situations, e.g.  a table-table join. See https://stackoverflow.com/questions/60123191/ksql-pull-query-on-a-joined-table for more info.

**Comment by derekjn** [comment]
> CREATE TABLE AS SELECT also needs this in some situations

Good point, I updated the description to account for CTAS.

**Comment by oscni** [comment]
any updates or roadmaps for this issue? Seems like a reoccurring requested feature.

Most of the examples around shows a standard use case of creating a table without aggregation functions, a stream and then joining them together. But what will actually happen internally in that case? will rockdbs kick in an materialize the table or will a new stream message search through the whole table(topic) from the beginning and try to find a match?

**Comment by mjsax** [comment]
It seems you are talking about a stream-table join? For this case, the table topic will be read and materialized into RocksDB, to allow stream-side records to do efficient lookups into the table state to compute the join result. However, this RocksDB state is _internal operator state_ and thus there is conceptually still no TABLE that you could issue pull queries against.

The main problem is really, that you could assume that you use the same table and to do two stream-table join queries using different input streams. For this case, each query will maintain its own RocksDB store. If one query is still catching up, its RocksDB state might be totally different to the other queries RocksDB state. And even if both queries process data at the "tail" of the input topics, both would be updated async and there is no guarantee that their RocksDB state is the same. Thus, if we would allow pull queries against this TABLE, which RocksDB should we use to answer the query? -- In an ideal world we would be able to share a single RocksDB for both queries, but this requires us to allow dynamic query rewriting what is not a trivial problem (because you might add a new query or stop an existing query that used the table at any point in time -- also for new queries, those would first work on older table state until they catch up and thus would need to maintain their own state initially, and would be able to reuse shared state only after they caught up, what make the dynamic query rewriting problem even harder).

Long story short: the feature is important and we are discussion about it in detail and hope to roll out a solution soon. However, it's more complicated as it might seem from a 10,000ft view and we might even roll out language changes to support it in a sane way. So stay tuned.

**Comment by oscni** [comment]
Ok I hoped that it was soon to be released now when kafka streams has the toTable method but maybe you're not using specifically that from streams and I guess it's always more to solve and think about when going into the details. Thanks for working on it, can't wait for it to start working :)
By the way, don't know if it's a separate issue but if/when you implement it into the REST API please also add some way to give only the result back in standard JSON format. Currently /query and /query-stream seems to output a header on the first line and after that the actual data but it would nice to have only the actual data outputted.

**Comment by mjsax** [comment]
> By the way, don't know if it's a separate issue but if/when you implement it into the REST API please also add some way to give only the result back in standard JSON format. Currently /query and /query-stream seems to output a header on the first line and after that the actual data but it would nice to have only the actual data outputted.

Feel free to open a new ticket for it.

**Comment by jbdoster** [comment]
It would be great to be able to derive 1 table from another aggregate table, this could make a work around to alias a key in a non-composite fashion

DML:

```
-- aggregates generate composite PK
CREATE TABLE T1 AS SELECT
    LATEST_BY_OFFSET(X) KEY,
    LATEST_BY_OFFSET(y) AS y_variant,
    LATEST_BY_OFFSET(z) AS z_variant,
    x, y, z
FROM
    DERIVED_STREAM_A
WHERE
    TYPE='topic._'
GROUP BY
    x, y, z;
    
   -- split and select the new PK
CREATE TABLE T2 AS SELECT
    SPLIT(KSQL_COL_0, '|+|')[0] KEY,
    KSQL_COL_0,
    y_variant,
    z_variant
FROM
    T1;
```
    
    Problem:
```
SELECT * FROM T2 WHERE KSQL_COL_0='x';
Can't pull from `T2` as it's not a materialized table. See https://cnfl.io/queries for more info.\nAdd EMIT CHANGES if you intended to issue a push query.
```

**Comment by cprasad1** [comment]
#7085 should close this
